pkg/kubelet/cm/dra/manager_test.go

```diff
@@ -0,0 +1,741 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/dynamic-resource-allocation/resourceclaim"
+	drapbv1 "k8s.io/kubelet/pkg/apis/dra/v1alpha2"
+	"k8s.io/kubernetes/pkg/kubelet/cm/dra/plugin"
+	"k8s.io/kubernetes/pkg/kubelet/cm/dra/state"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+)
+
+const (
+	driverName      = "test-cdi-device"
+	driverClassName = "test"
+)
+
+type fakeDRADriverGRPCServer struct {
+	drapbv1.UnimplementedNodeServer
+	driverName string
+	timeout    *time.Duration
+}
+
+func (s *fakeDRADriverGRPCServer) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+	if s.timeout != nil {
+		time.Sleep(*s.timeout)
+	}
+	deviceName := "claim-" + req.ClaimUid
+	result := s.driverName + "/" + driverClassName + "=" + deviceName
+	return &drapbv1.NodePrepareResourceResponse{CdiDevices: []string{result}}, nil
+}
+
+func (s *fakeDRADriverGRPCServer) NodeUnprepareResource(ctx context.Context, req *drapbv1.NodeUnprepareResourceRequest) (*drapbv1.NodeUnprepareResourceResponse, error) {
+	if s.timeout != nil {
+		time.Sleep(*s.timeout)
+	}
+	return &drapbv1.NodeUnprepareResourceResponse{}, nil
+}
+
+type tearDown func()
+
+func setupFakeDRADriverGRPCServer(t *testing.T, shouldTimeout bool) (tearDown, string) {
+	socketDir, err := os.MkdirTemp("", "dra")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	socketName := filepath.Join(socketDir, "server.sock")
+	stopCh := make(chan struct{})
+
+	teardown := func() {
+		close(stopCh)
+		os.RemoveAll(socketName)
+	}
+
+	l, err := net.Listen("unix", socketName)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	s := grpc.NewServer()
+	fakeDRADriverGRPCServer := &fakeDRADriverGRPCServer{
+		driverName: driverName,
+	}
+	if shouldTimeout {
+		timeout := plugin.PluginClientTimeout + time.Second
+		fakeDRADriverGRPCServer.timeout = &timeout
+	}
+
+	drapbv1.RegisterNodeServer(s, fakeDRADriverGRPCServer)
+
+	go func() {
+		go s.Serve(l)
+		<-stopCh
+		s.GracefulStop()
+	}()
+
+	return teardown, socketName
+}
+
+func TestNewManagerImpl(t *testing.T) {
+	kubeClient := fake.NewSimpleClientset()
+	for _, test := range []struct {
+		description        string
+		stateFileDirectory string
+		expectedError      bool
+	}{
+		{
+			description:        "invalid directory path",
+			stateFileDirectory: "",
+			expectedError:      true,
+		},
+		{
+			description:        "valid directory path",
+			stateFileDirectory: t.TempDir(),
+		},
+	} {
+		t.Run(test.description, func(t *testing.T) {
+			manager, err := NewManagerImpl(kubeClient, test.stateFileDirectory)
+			if test.expectedError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+				assert.NotNil(t, manager.cache)
+				assert.NotNil(t, manager.kubeClient)
+			}
+		})
+	}
+}
+
+func TestGetResources(t *testing.T) {
+	kubeClient := fake.NewSimpleClientset()
```

Comment:
most of the functions here need to use this fakeClient and NewManager(), can we make it a global var?

Reply:
I think we want to avoid global stuff

Reply:
then I think we can ignore what I said above